script/licenses.py

```diff
@@ -179,62 +179,77 @@ def from_dict(cls, data: PackageMetadata) -> PackageDefinition:
-        elif not approved and package.name not in EXCEPTIONS:
+    license_status = {
+        pkg.name: (pkg, check_license_status(pkg))
+        for data in raw_licenses
+        if (pkg := PackageDefinition.from_dict(data))
+    }
+
+    for name, version in TODO.items():
+        pkg, status = license_status.get(name, (None, None))
+        if pkg is None or not (version < pkg.version):
+            continue
+        assert status is not None
+
+        if status is True:
+            print(
+                f"Approved license detected for "
+                f"{pkg.name}@{pkg.version}: {get_license_str(pkg)}\n"
+                "Please remove the package from the TODO list.\n"
+            )
+        else:
             print(
                 "We could not detect an OSI-approved license for "
-                f"{package.name}@{package.version}: {package.license}"
+                f"{pkg.name}@{pkg.version}: {get_license_str(pkg)}\n"
+                "Please update the package version on the TODO list.\n"
             )
-            print()
-            exit_code = 1
-        elif approved and package.name in EXCEPTIONS:
+        exit_code = 1
+
+    for pkg, status in license_status.values():
+        if status is False and pkg.name not in EXCEPTIONS_AND_TODOS:
             print(
-                "Approved license detected for "
-                f"{package.name}@{package.version}: {package.license}"
+                "We could not detect an OSI-approved license for "
+                f"{pkg.name}@{pkg.version}: {get_license_str(pkg)}\n"
             )
-            print(f"Please remove the package from the EXCEPTIONS list: {package.name}")
-            print()
             exit_code = 1
-    current_packages = {package.name for package in package_definitions}
-    for package in [*TODO.keys(), *EXCEPTIONS]:
-        if package not in current_packages:
+        if status is True and pkg.name in EXCEPTIONS:
             print(
-                f"Package {package} is tracked, but not used. Please remove from the licenses.py"
-                "file."
+                f"Approved license detected for "
+                f"{pkg.name}@{pkg.version}: {get_license_str(pkg)}\n"
+                f"Please remove the package from the EXCEPTIONS list.\n"
             )
-            print()
             exit_code = 1
+
+    for name in EXCEPTIONS_AND_TODOS.difference(license_status):
+        print(
+            f"Package {name} is tracked, but not used. "
+            "Please remove it from the licenses.py file.\n"
+        )
+        exit_code = 1
+
     return exit_code
 
 
+def check_license_status(package: PackageDefinition) -> bool:
+    """Check if package licenses is OSI approved."""
+    for approved_license in OSI_APPROVED_LICENSES:
+        if approved_license in package.license:
+            return True
+    return False
+
+
+def get_license_str(package: PackageDefinition) -> str:
+    """Return license string."""
+    return f"{package.license}"
```

Comment:
I'm assuming we're planning to change this logic in the future. If not, this function isn't really needed.

Reply:
Yes, I plan to separate the license sting into each component (expression, metadata, classifier) in a followup PR.